### DISCUSSION TITLE
diffusion gradient nonlinearity correction

### DISCUSSION THREAD

**Discussion by dcdeaniii** [dipy/dipy#3117] [Reconstruction]
Hi - we're interested in correcting the bmatrix for diffusion gradient nonlinearities, where the bvals/bvecs vary spatially. We can compute the gradient deviation tensor map from our scanner's gradient nonlinearity coefficients.  Is there a way to generate a voxelwise GradientTable that can then be used in the fitting of different reconstruction models (e.g. DTI)? Currently, we can apply this correction in the model fit by looping over all voxels in our image, however, the fitting procedure is pretty inefficient.

**Comment by arokem** [comment]
We don't have something like that currently implemented, so looping over voxels is your best bet for now. 

Out of curiosity: if you implemented that as part of the GradientTable object, how would you make this more efficient than a loop over voxels? I think we could implement this to avoid duplication of code in user's code and to make this kind of use-case more convenient, but I am not sure that I see how we would me it more _efficient_.

**Comment by samcoveney** [comment]
I have been thinking about the same thing - honestly, I cannot see a better solution than loop over voxels. The implementation of iterations like in `dki.py` using `multi_voxel` would probably be the way to go, with an option to pass in a different `design_matrix` on each iteration...
